Package validation: accept packages of size 1

instagibbs opened the pull request:
There's no particular reason to restrict single transaction submissions with submitpackage. This change relaxes the RPC checks as enables the `AcceptPackage` flow to accept packages of a single transaction. 

Resolves #31085

TheCharlatan commented:
Cobcept ACK

stickies-v commented on src/validation.cpp:
```diff
@@ -1695,15 +1695,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
         return PackageMempoolAcceptResult(package_state_quit_early, {});
     }
 
-    // IsChildWithParents() guarantees the package is > 1 transactions.
-    assert(package.size() > 1);
```
I wouldn't remove this, but rather update to:

```
    // IsChildWithParents() guarantees the package is not empty.
    assert(!package.empty());
```

After all, if the package is empty, we'd have UB a few lines down.

stickies-v commented on src/validation.cpp:
```diff
@@ -1695,15 +1695,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
-    std::transform(package.cbegin(), package.cend() - 1,
-                   std::inserter(unconfirmed_parent_txids, unconfirmed_parent_txids.end()),
-                   [](const auto& tx) { return tx->GetHash(); });
+    if (package.size() > 1) {
```
I think this check can be dropped, operation should be safe and no-op if package is empty? I'm fine with it either way, just simplifies the code a bit.

instagibbs replied:
I'm bad at reading the spec, is that defined behavior? https://en.cppreference.com/w/cpp/algorithm/transform

instagibbs replied:
ok I'm convinced, removed

stickies-v reviewed (commented):
Concept ACK

glozow commented on src/validation.cpp:
```diff
@@ -1695,15 +1695,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
         return PackageMempoolAcceptResult(package_state_quit_early, {});
     }
 
-    // IsChildWithParents() guarantees the package is > 1 transactions.
-    assert(package.size() > 1);
```
Given that this assert exists, I'm worried we have other places with very strict assumptions on what `IsChildWithParents` can guarantee. Perhaps it's safer to just have the RPC code or `ProcessNewPackage` to direct a single transaction to ATMP instead.

instagibbs replied:
I'd rather not; we almost immediately handle singletons via `AcceptSubPackage` internally, I'm not convinced this is more risky than adding more complexity on top.

glozow replied:
This is what testmempoolaccept does

glozow reviewed (commented):
concept ACK

stickies-v commented on test/functional/rpc_packages.py:
```diff
@@ -381,13 +381,15 @@ def test_submitpackage(self):
             node.submitpackage, [chain_hex[0]] * (MAX_PACKAGE_COUNT + 1)
         )
 
+        # Singleton is ok
+        node.submitpackage([chain_hex[0]] * 1)
+
```
No need to special-case this:

<details>
<summary>git diff on 72872c7073</summary>

```diff
diff --git a/test/functional/rpc_packages.py b/test/functional/rpc_packages.py
index d4da0f9b06..ce42bf810e 100755
--- a/test/functional/rpc_packages.py
+++ b/test/functional/rpc_packages.py
@@ -370,7 +370,7 @@ class RPCPackagesTest(BitcoinTestFramework):
         node = self.nodes[0]
 
         self.log.info("Submitpackage valid packages with 1 child and some number of parents")
-        for num_parents in [1, 2, 24]:
+        for num_parents in [0, 1, 2, 24]:
             self.test_submit_child_with_parents(num_parents, False)
             self.test_submit_child_with_parents(num_parents, True)
 
@@ -387,9 +387,6 @@ class RPCPackagesTest(BitcoinTestFramework):
             node.submitpackage, [chain_hex[0]] * (MAX_PACKAGE_COUNT + 1)
         )
 
-        # Singleton is ok
-        node.submitpackage([chain_hex[0]] * 1)
-
         # Create a transaction chain such as only the parent gets accepted (by making the child's
         # version non-standard). Make sure the parent does get broadcast.
         self.log.info("If a package is partially submitted, transactions included in mempool get broadcast")

```
</details>

instagibbs replied:
done

stickies-v commented on src/test/txpackage_tests.cpp:
```diff
@@ -283,6 +283,14 @@ BOOST_AUTO_TEST_CASE(noncontextual_package_tests)
         BOOST_CHECK(GetPackageHash({tx_child, tx_parent}) != GetPackageHash({tx_child, tx_child}));
+
+        // IsChildWithParents* can also be a singleton
+        BOOST_CHECK(IsChildWithParents({tx_parent}));
+        BOOST_CHECK(IsChildWithParentsTree({tx_parent}));
```
nit: we require exactly one child, so I think using `tx_parent` is unnecessarily confusing
```suggestion
        BOOST_CHECK(IsChildWithParents({tx_child}));
        BOOST_CHECK(IsChildWithParentsTree({tx_child}));
```

instagibbs replied:
sure done

stickies-v commented on src/policy/packages.cpp:
```diff
@@ -118,8 +118,9 @@ bool IsWellFormedPackage(const Package& txns, PackageValidationState& state, boo
+    if (package.empty()) return false;
     assert(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}));
-    if (package.size() < 2) return false;
+    if (package.size() < 2) return true;
```
Note: this also requires updating this documentation: https://github.com/bitcoin/bitcoin/blob/538ccaed004ff89ec8b2b71d8711feed624ffccc/doc/policy/packages.md?plain=1#L78-L79

rkrux replied:
Worth updating RPC doc as well here? Informing that only child is also allowed.

https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mempool.cpp#L926

stickies-v replied:
I had similar thoughts, yes. I think the current description is correct but less helpful than it could be (even without this PR).

Suggested alternative:

"The package must solely consist of a child transaction and all, some or none of its unconfirmed parents. None of the parents may depend on each other.\n"

instagibbs replied:
took a slightly different tack, I think this is more correct. You *must* include unconfirmed parents, if any exist.

stickies-v replied:
> You _must_ include unconfirmed parents, if any exist.

Thanks, you're right. I was confused by the `IsChildWithParents` documentation stating:

> ...not all parents need to be present...

But that is of course just an incomplete, context-free check. The `MemPoolAccept::AcceptPackage` documentation and implementation indeed require all unconfirmed parents to be included. Your [suggestion](https://github.com/bitcoin/bitcoin/compare/72872c7073e99da0828e788b25520dcecf721c1b..92ed1bd586f7048f75b38024521696ff04e5f69c) seems better.

Perhaps including the no-parents case in the RPC example would be helpful too:

<details>
<summary>git diff on 92ed1bd586</summary>

```diff
diff --git a/src/rpc/mempool.cpp b/src/rpc/mempool.cpp
index 09a9d87c2c..3074b3994a 100644
--- a/src/rpc/mempool.cpp
+++ b/src/rpc/mempool.cpp
@@ -965,8 +965,8 @@ static RPCHelpMan submitpackage()
             },
         },
         RPCExamples{
-            HelpExampleRpc("submitpackage", R"(["rawtx1", "rawtx2"])") +
-            HelpExampleCli("submitpackage", R"('["rawtx1", "rawtx2"]')")
+            HelpExampleRpc("submitpackage", R"(["raw-parent-tx-1", "raw-parent-tx-2", "raw-child-tx"])") +
+            HelpExampleCli("submitpackage", R"('["raw-tx-without-unconfirmed-parents"]')")
         },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
         {

```
</details>

instagibbs replied:
can do when touched again

instagibbs replied:
done

stickies-v reviewed (commented):
Approach ACK, code LGTM 72872c7073e99da0828e788b25520dcecf721c1b and change looks safe at first sight but want to take some more time to think about it as I'm not super familiar with this code.

kevkevinpal commented:
Concept ACK

rkrux reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/commit/72872c7073e99da0828e788b25520dcecf721c1b
Would be great to have the docs updated as well.

Make and functional tests successful.

In regtest I manually created a PSBT, processed it and submitted a single transaction package with successful response.

```
➜  bitcoin git:(2024-10-submitpackage-singleton) ✗ bitcoincli submitpackage "[\"020000000001018e61beaa1f85daba48a46e6bef4e58d12f32768d347185cb077c988cc5af976b0000000000fdffffff02e60f1024010000001600148d6b22f6e4c0c5dc96529ca5ebc6ad0cde42966800e1f5050000000016001468c93e48eeed6fda96d6d234da5c3ccb5e052856024730440220292c5b776a8d1280bced38c1ef463ff204bd815801798e16909bd9dfd20e495a02206a0d2c1a55bad94b69411ca9ed4656d506066a5aa88265bfd7a08a644bc015840121039814ceadf23fc32f60fa4f607c4882f20e5cb12f766c66407dbc61f0cc51e30d00000000\"]"
{
  "package_msg": "success",
  "tx-results": {
    "6f3a4de5d782679609abd424dcfd3711472c8fdb5270695a3c0636bad4717e09": {
      "txid": "108cd63d68165c7af07cf410a6055eade488cb3d049935c79832f8b96f817e69",
      "vsize": 141,
      "fees": {
        "base": 0.00000282,
        "effective-feerate": 0.00002000,
        "effective-includes": [
          "6f3a4de5d782679609abd424dcfd3711472c8fdb5270695a3c0636bad4717e09"
        ]
      }
    }
  },
  "replaced-transactions": [
  ]
}
```

instagibbs commented:
@glozow we may want to consider relaxing `package-not-child-with-unconfirmed-parents` to allow people to submit just the cpfp alone? Should be done in a separate PR regardless.

glozow commented:
> @glozow we may want to consider relaxing `package-not-child-with-unconfirmed-parents` to allow people to submit just the cpfp alone? Should be done in a separate PR regardless.

That works without validation changes if you make the split happen within RPC or ProcessNewPackage.

sedited reviewed (approved):
lgtm ACK 92ed1bd586f7048f75b38024521696ff04e5f69c

Though @stickies-v's [suggestion](https://github.com/bitcoin/bitcoin/pull/31096#discussion_r1809122657) would be a nice addition. This was non-obvious to me.

stickies-v reviewed (approved):
ACK 303661871debadee5f67bd7e4cd0cccc85344ef2

I think reducing the special (and imo unexpected) treatment of packages of transactions without unconfirmed parents by this PR is an improvement. I [looked](https://github.com/bitcoin/bitcoin/pull/31096#discussion_r1810732642) at the call hierarchy of `IsChildWithParents()`, and the changes look safe to me, but I am not super familiar with this part of the code.

stickies-v commented:
re-ACK 303661871debadee5f67bd7e4cd0cccc85344ef2, thanks for adding the test.

naumenkogs commented on src/validation.cpp:
```diff
@@ -1695,8 +1695,9 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
     }
 
-    // IsChildWithParents() guarantees the package is > 1 transactions.
-    assert(package.size() > 1);
```
Won't we redundantly call `AcceptSubPackage` twice for a single transaction, if `individual_results_nonfinal.emplace(wtxid, single_res);` stuff is triggered?

instagibbs replied:
you're right, relatively harmless, but pushed a change which means it won't try package evaluation unless `txns_package_eval` is larger than size 1.

instagibbs replied:
It also causes double-validation for single transaction package that is missing inputs.

Could special case this but not sure I love making things even harder to understand:
```
diff --git a/src/validation.cpp b/src/validation.cpp
index 3e48335255..c89050e895 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1790,10 +1790,16 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
                 // some of them may still be valid.
                 quit_early = true;
                 package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
                 individual_results_nonfinal.emplace(wtxid, single_res);
             } else {
+                // It's retryable in a package but this is a singleton so we should mark the package failed
+                if (package.size() == 1) {
+                    quit_early = true;
+                    package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+                }
+
                 individual_results_nonfinal.emplace(wtxid, single_res);
                 txns_package_eval.push_back(tx);
             }
         }
     }
```

thoughts?

naumenkogs replied:
I've failed to suggest a simple code-change to avoid double-validation of a single tx.

The code is really about multi-tx. Even if we add a bunch of conditions around `auto multi_submission_result`, stuff below (`individual_results_nonfinal`) makes you think about multi-tx again.

The least thing I'm worried about is double-validation on itself.
I'm worried about code comprehension complexity: reading multi-tx code against a single-tx scenario.

At this point I agree with Gloria to handle this case earlier. Or substantially moving code around inside this function to make reasoning about 1-conf more natural to it.

naumenkogs replied:
The fix for double-validation seems good.
--------------
Do you mind sharing the code of your attempt (assuming it compiles :))? I wanna take a look, but would be useful for archival reasons anyway.

instagibbs replied:
> Do you mind sharing the code of your attempt (assuming it compiles :))? I wanna take a look, but would be useful for archival reasons anyway.

It compiled, but it won't pass tests because I would have had to heavily modify the BroadcastTransaction interface. Not sure I kept the branch sorry.

naumenkogs replied:
So you'd rather not take `else if (package.size() == 1 || //` still? I think it's an improvement for code comprehension if we end up handling single-tx here.

naumenkogs replied:
Expanding on this thought, I sketched a two-step improvement in handling single-package tx wrt code readability. Wanna [take a look](https://github.com/naumenkogs/bitcoin/commits/2024-10-submitpackage-singleton-v1/)?

I certainly might be missing something.

instagibbs replied:
Left a comment, I think my one-liner is the cleanest correct solution?

pushed it as a separate commit

naumenkogs replied:
Have you left a comment on my branch? I can't see it for some reason.

instagibbs replied:
https://github.com/naumenkogs/bitcoin/commit/88c10c5789bfae1865d03966d1ce523e870ff0bc#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1811 weird comment UI, here is a direct link hopefully

naumenkogs replied:
Weird UI indeed, i left a reply and I'll try to monitor that thread now.

instagibbs commented:
rebased on master due to test conflict

stickies-v commented:
re-ACK 5e4df9f67179e9cc284cb2ad2264de6b8bb6c606, no changes except for addressing merge conflict from #31139

naumenkogs commented:
@stickies-v i think you might be looking at the wrong commit. The hash you acked is outdated.

sedited reviewed (approved):
Re-ACK d120154d55451b6c90b709a4a093cbb6cabe48d8

Added improved rpc help and unit test case since last review.

glozow committed:
rpc: Allow single transaction through submitpackage

And under the hood suppoert single transactions
in AcceptPackage. This simplifies user experience
and paves the way for reducing number of codepaths
for transaction acceptance in the future.

Co-Authored-By: instagibbs <gsanders87@gmail.com>

glozow reviewed (commented):
I still don't really like that the `IsChildWithParents()` function returns true for a tx that isn't a child / doesn't have any parents. Also dislike that we run the whole input-loading thing for the child-with-unconfirmed-parents check even for a singleton tx. I guess one way around this is to rename the 2 `IsChildWithParents` functions to something along the lines of "is topology we can handle" and to add a `package.size() > 1` gate to the child-with-unconfirmed-parents check.

I spent some time implementing the redirection at `submitpackage` and at `ProcessNewPackage`, and I now agree with you that it's not as clean. Considering `AcceptSubPackage` is already smoothing out those joints, I like the idea of getting rid of ATMP in the future, just throwing things at `ProcessNewPackage`, and enumerating what we can handle at the start of `AcceptPackage`.

Here is my counteroffer [branch](https://github.com/glozow/bitcoin/commits/2024-11-acceptpackage-single/) which is fairly similar to this. I've also implemented a removal of the is-child-with-unconfirmed-parents rule on top of it, which I think we're ready for after this PR. There's also an extra unit test in between. Lmk what you think?

instagibbs commented:
@glozow I think I agree your branch at 4737df3f6512d2d9c7f8aa95e0635b9d03031402 is superior to this. 

>  Considering AcceptSubPackage is already smoothing out those joints, I like the idea of getting rid of ATMP in the future, just throwing things at ProcessNewPackage, and enumerating what we can handle at the start of AcceptPackage.

Yes, I would love to get rid of the separate paths over time.

> 've also implemented a removal of the is-child-with-unconfirmed-parents rule on top of it, which I think we're ready for after this PR

I'll have to do a bunch more thinking and clearly requires its own PR.

glozow commented:
thanks for looking!

> I'll have to do a bunch more thinking and clearly requires its own PR.
>
In case it wasn’t clear, I definitely meant that as a separate PR. I only
implemented it to sketch out what it’d look like as the next step.

>

instagibbs commented:
Took @glozow first two commits with a couple additional sanity check lines.

glozow reviewed (commented):
> edit: Turns out it doesn't correctly mark package as failed if single tx fails.

Sorry about that! I'm happy with the `package.size() == 1` gate for the retry logic. Nice that the fuzzer caught that. Can add a low fee test case to package_single_tx to hit it I think.

ACK 278cb41e8785a4948972a4b4f635c2654edd0e3e

naumenkogs commented:
ACK 278cb41e8785a4948972a4b4f635c2654edd0e3e

instagibbs commented:
Would have been remiss to not push a regression test for the issue the fuzzer caught. Pushed test case for coverage of a reconsiderable tx.

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -492,6 +494,93 @@ BOOST_AUTO_TEST_CASE(package_submission_tests)
+    BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
+    if (auto err_single{CheckPackageMempoolAcceptResult(package_tx_single_low_fee, result_single_tx_low_fee, /*expect_valid=*/false, m_node.mempool.get())}) {
+        BOOST_ERROR(err_single.value());
+    }
```
Check the error?
```
BOOST_CHECK_EQUAL(result_single_tx_low_fee.m_state.GetResult(), PackageValidationResult::PCKG_TX);
auto it_low_fee = result_single_tx_low_fee.m_tx_results.find(tx_single_low_fee->GetWitnessHash());
BOOST_CHECK_EQUAL(it_low_fee->second.m_state.GetResult(), TxValidationResult::TX_RECONSIDERABLE);
```

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -492,6 +494,97 @@ BOOST_AUTO_TEST_CASE(package_submission_tests)
+    BOOST_CHECK_EQUAL(result_single_tx_low_fee.m_state.GetResult(), PackageValidationResult::PCKG_TX);
+    auto it_low_fee = result_single_tx_low_fee.m_tx_results.find(tx_single_low_fee->GetWitnessHash());
+    BOOST_CHECK_EQUAL(it_low_fee->second.m_state.GetResult(), TxValidationResult::TX_RECONSIDERABLE);
+    if (auto err_single{CheckPackageMempoolAcceptResult(package_tx_single_low_fee, result_single_tx_low_fee, /*expect_valid=*/false, m_node.mempool.get())}) {
```
nit: in the future, best to call `CheckPackageMempoolAcceptResult` beforehand because the errors would be more helpful if they happened (e.g. "error this value should exist" instead of bad optional access)

glozow reviewed (commented):
ACK 32fc59796f74a2941772b5ec2755b1319132cd9c

naumenkogs commented:
ACK https://github.com/bitcoin/bitcoin/commit/32fc59796f74a2941772b5ec2755b1319132cd9c

achow101 commented:
ACK 32fc59796f74a2941772b5ec2755b1319132cd9c
